libbpf-tools/bashreadline.c

```diff
@@ -0,0 +1,232 @@
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 's':
+		libreadline_path = strdup(arg);
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_size)
+{
+	struct str_t *e = data;
+	struct tm *tm;
+	char ts[16];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%m:%S", tm);
+
+	printf("%-9s %-6lld %s\n", ts, e->pid, e->str);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+static char *find_readline_so() {
+	const char *bash_path = "/bin/bash";
+	FILE *fp;
+	char *line = NULL;
+	char *sep, *path_start, *sp;
+	size_t line_sz;
+	off_t func_off;
+	char *result = NULL;
+
+	func_off = get_elf_func_offset(bash_path, "readline");
+	if (func_off >= 0) {
+		return strdup(bash_path);
+	}
+
+	/*
+	 * Try to find libreadline.so if the symbol of readline is not
+	 * defined in bash
+	 */
+	fp = popen("ldd /bin/bash", "r");
+	if (fp == NULL)
+		goto cleanup;
+	while (getline(&line, &line_sz, fp) >= 0) {
+		if (strstr(line, "libreadline.so") == NULL)
+			continue;
+		sep = strstr(line, " => ");
+		if (sep == NULL)
+			goto cleanup;
+		path_start = sep + 4;
+		if (isspace(*path_start))
+			goto cleanup;
+		sp = strstr(path_start, " ");
+		if (sp == NULL)
+			goto cleanup;
+		path_start[sp - path_start] = 0;
+		if (strstr(path_start, "/libreadline.so") == NULL)
+			goto cleanup;
+		result = strdup(path_start);
+		break;
+	}
+
+cleanup:
+	if (line)
+		free(line);
+	return result;
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct bashreadline_bpf *obj = NULL;
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct bpf_link *link = NULL;
+	char *readline_so_path = NULL;
+	off_t func_off;
+	int err;
+	int fd;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	if (libreadline_path) {
+		readline_so_path = libreadline_path;
+	} else if ((readline_so_path = find_readline_so()) == NULL) {
+		warn("failed to find readline\n");
+		return 1;
+	}
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %d\n", err);
+		goto cleanup;
+	}
+
+	obj = bashreadline_bpf__open();
+	if (!obj) {
+		warn("failed to open BPF object\n");
+		goto cleanup;
+	}
+
+	err = bashreadline_bpf__load(obj);
+	if (err) {
+		warn("failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	func_off = get_elf_func_offset(readline_so_path, "readline");
+	if (func_off < 0) {
+		warn("cound not find readline in %s\n", readline_so_path);
+		goto cleanup;
+	}
+
+	link = bpf_program__attach_uprobe(obj->progs.printret, true, -1,
```

Comment:
it's best to set it to `obj->links.printret = bpf_program__attach_uprobe(...)` so that skeleton's destroy() will automatically destroy the link as well. Simpler clean up that way.